Mage.Sets/src/mage/cards/c/CallForAid.java

```diff
@@ -0,0 +1,127 @@
+import mage.abilities.keyword.HasteAbility;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.CardType;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.filter.common.FilterCreaturePermanent;
+import mage.filter.predicate.permanent.PermanentInListPredicate;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.target.common.TargetOpponent;
+
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * @author spillnerdev
+ */
+public final class CallForAid extends CardImpl {
+
+    public CallForAid(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.SORCERY}, "{4}{R}");
+
+        // Gain control of all creatures target opponent controls until end of turn. Untap those creatures. They gain haste until end of turn. You can't attack that player this turn. You can't sacrifice those creatures this turn.
+        getSpellAbility().addTarget(new TargetOpponent());
+        getSpellAbility().addEffect(new CallForAidEffect());
+    }
+
+    private CallForAid(final CallForAid card) {
+        super(card);
+    }
+
+    @Override
+    public CallForAid copy() {
+        return new CallForAid(this);
+    }
+}
+
+
+class CallForAidEffect extends OneShotEffect {
+
+    public CallForAidEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Gain control of all creatures target opponent controls until end of turn. Untap those creatures. They gain haste until end of turn. You can't attack that player this turn. You can't sacrifice those creatures this turn.";
+    }
+
+    protected CallForAidEffect(CallForAidEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        UUID targetOpponent = getTargetPointer().getFirst(game, source);
+        if (targetOpponent != null) {
+            // First we get all creatures of target player...
+            FilterCreaturePermanent filter = new FilterCreaturePermanent();
+            List<Permanent> opponentCreatures = game.getBattlefield().getActivePermanents(filter, targetOpponent, game);
+            filter.add(new PermanentInListPredicate(opponentCreatures));
+            //... gain control of them till end of turn...
+            GainControlAllEffect gainControl = new GainControlAllEffect(Duration.EndOfTurn, filter);
+            gainControl.apply(game, source);
+            //...untap all stolen creatures...
+            new UntapAllEffect(filter).apply(game,source);
+            //...give them haste...
+            GainAbilityAllEffect gainHaste = new GainAbilityAllEffect(HasteAbility.getInstance(), Duration.EndOfTurn, filter);
+            game.addEffect(gainHaste, source);
+            //...make sure they can't attack their owner...
+            GainAbilityAllEffect cantAttackOwnerEffect = new GainAbilityAllEffect(new SimpleStaticAbility(new CantAttackOwnerEffect()), Duration.EndOfTurn, filter);
```

Comment:
This is not correct. The text of the card is "You can't attack that player this turn." It is not related to the creatures nor their owners.

Reply:
First of all, thanks for the extensive feedback. I reworked the code. However this "can't attack that player this turn" gave me a bit of a headache and I am certain the current implementation, while working, might cause issues down the line.

I'd be happy if you could take a look at it and tell me if it's acceptable. In the meantime I will continue to wrap my head around the engine and eventually rework the effect once my understanding of effect application and resolution increases.